### ISSUE TITLE
Implement asteroids

### ISSUE THREAD

**Issue by johnthagen** [johnthagen/rust-belt#7]
- [x] Asteroids should appear randomly in the window
- [x] Asteroids drift in one direction with a random starting velocity
- [x] Jagged asteroids (started by @mcdenhoed)
- [x] Spawn asteroids with a random radius within a certain range (small and large Asteroids).
- [x] ~~Don't spawn asteroid directly with some distance of `Player`~~ Spawn asteroids from the edge of the screen. #84
- [x] Track total time in game and increase asteroid spawn rate based upon that #84

See https://crates.io/crates/rand

See 
- https://github.com/aochagavia/rocket/blob/71c14c94f34df8b4c7a086ae0a1212b3e3e3f150/src/drawing/point.rs#L19
- https://github.com/aochagavia/rocket/blob/71c14c94f34df8b4c7a086ae0a1212b3e3e3f150/src/models/enemy.rs#L15

**Comment by mcdenhoed** [comment]
@johnthagen @theandrewdavis See https://github.com/johnthagen/rust-belt/tree/asteroids
I figured that a good base step would be to create a circle and mutate it. I've done the circle part.

**Comment by johnthagen** [comment]
Cool! Just so I understand correctly, are you using a `polygon` so that eventually it can be made "asteroid-shaped" rather than purely circular (using `ellipse`)?

**Comment by mcdenhoed** [comment]
Correct!

**Comment by mcdenhoed** [comment]
Also, this way, the asteroid shape can be randomly generated on creation.

**Comment by mcdenhoed** [comment]
@johnthagen Asteroids are now non-circular!

**Comment by johnthagen** [comment]
:star2:!

**Comment by johnthagen** [comment]
@mcdenhoed 

- [x] `[[f64;2]; NUM_SEGMENTS]` is repeated a lot. Should we use a [type alias](https://doc.rust-lang.org/book/type-aliases.html) to something like `CircularPolygon`?
 - It starts out as a ["Regular Polygon"](https://en.wikipedia.org/wiki/Regular_polygon) and then is mutated if we want to use the more technical term. Basically we want to capture the fact that it's a polygon that is generally "circle-looking" with some mutation.
- Another name could be ["Circumscribed Circle"](https://en.wikipedia.org/wiki/Circumscribed_circle) though I think "circumcircle" sounds fun :)

**Comment by mcdenhoed** [comment]
@johnthagen I think that's a reasonable course of action.

**Comment by johnthagen** [comment]
- [x] ~~Feels a little strange for `randomize_shape()` to both take a mutable reference to `shape` and return `shape`. Seems like it could be a simpler API to just mutate it, and let the caller use the original mutated shape?~~ It's not taking a mutable reference (`&mut`), it's taking a copy and then mutating that copy. It's a pure function this way, so I think it's fine to leave.

- [x] `let new_shape : [[f64;2];NUM_SEGMENTS] = generate_circle(radius);` probably doesn't need the type hint.

**Comment by mcdenhoed** [comment]
@johnthagen Agreed

**Comment by johnthagen** [comment]
#69 Implements part of this.

**Comment by mcdenhoed** [comment]
@johnthagen More changes here: https://github.com/johnthagen/rust-belt/tree/asteroid-spawning

**Comment by mcdenhoed** [comment]
Could maybe center the asteroid spawn radius around the player instead of window center? Not sure what would be best.

**Comment by mcdenhoed** [comment]
Other idea: maybe we should keep track of DT and have the spawn be time based? And just slowly increase the rate?

**Comment by mcdenhoed** [comment]
Made the asteroids time based. It plays a lot better now. Still need to add them breaking up.

**Comment by johnthagen** [comment]
Wow that plays so much better!

**Comment by johnthagen** [comment]
I think some docstrings explaining how/why the Asteroid methods and functions work would be really helpful, since the connection to how things are drawn on the screen isn't always straightforward from just reading the code. "Why" comments through the code would be great too since there is a lot of math going on. Basically #73.

A couple snippets that aren't clear to me at first:

```rust
let radius = cmp::max(window_size.width, window_size.height) as f64 + RADIUS;
```

```rust
        if self.on_screen == false {
            if self.pos.x > RADIUS && self.pos.x + RADIUS < self.window_size.width as f64 / 2.0 &&
               self.pos.y > RADIUS &&
               self.pos.y + RADIUS < self.window_size.height as f64 / 2.0 {
                self.window_size.width /= 2;
                self.window_size.height /= 2;
                self.on_screen = true;
            }
        }
```

I looks like asteroids always are directed at the `Player`. I wonder if that is necessary. It would be a bit simpler if they just starting flying in randomly, and as they field density increases it will still provide a challenge. It sort of feels like an invisible force is "shooting" asteroids at me rather than moving through a random debris field.

I'd like to postpone breaking the asteroids up (#63) until after 1.0 to keep 1.0 a minimal viable game. I think this branch is very close to being complete and then 1.0 only needs to track score #11 before being feature complete.

**Comment by johnthagen** [comment]
```rust
use std::f64;
...
const PI: f64 = f64::consts::PI;
```

Can be simplified to:

```rust
use std::f64::consts::PI;
```

I agree we should just use it as `PI` though, since we use it so many times.

**Comment by mcdenhoed** [comment]
@johnthagen Are we doing random sizes in version 1? Is that even likely to hang around once we get splitting?

**Comment by johnthagen** [comment]
I think we should do both. I think it would still hang around even with splitting. Some would just start smaller. I think it would make it feel a little more dynamic and random than always having the same big asteroid spawn in.

So random radius in version 1.0.

And then random radius + splitting in 2.0.

**Comment by mcdenhoed** [comment]
Implemented random radius in #91
